spark/v3.5/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestRowLineagePropagation.java

```diff
@@ -0,0 +1,441 @@
+    return records;
+  }
+
+  /**
+   * Generates numRecords records with data values as 0 to numRecords - 1, with row lineage, each
+   * record is partitioned based on the provided partition generator. The row ID and last updated
+   * sequence number assigned is the same as the data value for that row.
+   *
+   * @return a partitioned map
+   */
+  protected PartitionMap<List<Record>> createRecordsWithRowLineage(
+      Schema schema,
+      PartitionSpec spec,
+      int numRecords,
+      Function<Record, StructLike> partitionGenerator) {
+    PartitionMap<List<Record>> recordsByPartition =
+        PartitionMap.create(Map.of(spec.specId(), spec));
+    for (int i = 0; i < numRecords; i++) {
+      long rowId = i;
+      long lastUpdatedSequenceNumber = i;
+      Record record = createRecordWithRowLineage(schema, i, rowId, lastUpdatedSequenceNumber);
+      StructLike partition = partitionGenerator != null ? partitionGenerator.apply(record) : null;
+      List<Record> recordsForPartition = recordsByPartition.get(spec.specId(), partition);
+      if (recordsForPartition == null) {
+        recordsForPartition = Lists.newArrayList();
+      }
+
+      recordsForPartition.add(record);
+      recordsByPartition.put(spec.specId(), partition, recordsForPartition);
+    }
+
+    return recordsByPartition;
+  }
+
+  protected List<Record> filterByDataNotEqualTo(List<Record> records, int data) {
+    return records.stream()
+        .filter(record -> record.get(0, Integer.class) != data)
+        .collect(Collectors.toList());
+  }
+
+  protected void appendRecords(Table table, PartitionMap<List<Record>> partitionedRecords)
+      throws IOException {
+    AppendFiles append = table.newAppend();
+
+    for (Map.Entry<Pair<Integer, StructLike>, List<Record>> entry : partitionedRecords.entrySet()) {
+      OutputFile file = Files.localOutput(temp.resolve(UUID.randomUUID().toString()).toFile());
+      DataWriter<Record> writer =
+          new GenericAppenderFactory(schemaWithRowLineage(table), table.spec())
+              .newDataWriter(
+                  EncryptionUtil.plainAsEncryptedOutput(file), fileFormat, entry.getKey().second());
+      List<Record> recordsForPartition = entry.getValue();
+      writer.write(recordsForPartition);
+      writer.close();
+      append =
+          append
+              .appendFile(writer.toDataFile())
+              .toBranch(branch != null ? branch : SnapshotRef.MAIN_BRANCH);
+    }
+
+    append.commit();
+  }
+
+  protected void assertRecordsAreEqual(List<Record> expected, List<Record> actual) {
+    assertThat(actual.size()).isEqualTo(expected.size());
+    for (int i = 0; i < expected.size(); i++) {
+      Record expectedRecord = expected.get(i);
+      Record actualRecord = actual.get(i);
+      assertThat(expectedRecord.get(0)).isEqualTo(actualRecord.get(0));
+      assertThat(expectedRecord.get(1)).isEqualTo(actualRecord.get(1));
+      assertThat(expectedRecord.get(2)).isEqualTo(actualRecord.get(2));
```

Comment:
in MoR the records returned by the generics code includes the file position for pos deletes so an exact equality faills since there's an additional field...that's why this compares the data, row id, sequence  values directly. I'll clean this up so it's more obvious to a reader